memory/table_data.go

```diff
@@ -0,0 +1,320 @@
+func (td *TableData) errIfDuplicateEntryExist(cols []string, idxName string) error {
+	columnMapping, err := td.columnIndexes(cols)
+	if err != nil {
+		return err
+	}
+	unique := make(map[uint64]struct{})
+	for _, partition := range td.partitions {
+		for _, row := range partition {
+			idxPrefixKey := projectOnRow(columnMapping, row)
+			if hasNulls(idxPrefixKey) {
+				continue
+			}
+			h, err := sql.HashOf(idxPrefixKey)
+			if err != nil {
+				return err
+			}
+			if _, ok := unique[h]; ok {
+				return sql.NewUniqueKeyErr(formatRow(row, columnMapping), false, nil)
+			}
+			unique[h] = struct{}{}
+		}
+	}
+	return nil
+}
+
+func hasNulls(row sql.Row) bool {
+	for _, v := range row {
+		if v == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// getField returns the index and column index with the name given, if it exists, or -1, nil otherwise.
+func (td *TableData) getField(col string) (int, *sql.Column) {
+	i := td.schema.IndexOf(col, td.tableName)
+	if i == -1 {
+		return -1, nil
+	}
+
+	return i, td.schema.Schema[i]
+}
+
+func (td *TableData) generateCheckName() string {
+	i := 1
+Top:
+	for {
+		name := fmt.Sprintf("%s_chk_%d", td.tableName, i)
+		for _, check := range td.checks {
+			if check.Name == name {
+				i++
+				continue Top
+			}
+		}
+		return name
+	}
+}
+
+func (td *TableData) indexColsForTableEditor() ([][]int, [][]uint16) {
+	var uniqIdxCols [][]int
+	var prefixLengths [][]uint16
+	for _, idx := range td.indexes {
+		if !idx.IsUnique() {
+			continue
+		}
+		var colNames []string
+		expressions := idx.(*Index).Exprs
+		for _, exp := range expressions {
+			colNames = append(colNames, exp.(*expression.GetField).Name())
+		}
+		colIdxs, err := td.columnIndexes(colNames)
+		if err != nil {
+			// this means that the column names in this index aren't in the schema, which can happen in the case of a
+			// table rewrite
+			continue
+		}
+		uniqIdxCols = append(uniqIdxCols, colIdxs)
+		prefixLengths = append(prefixLengths, idx.PrefixLengths())
+	}
+	return uniqIdxCols, prefixLengths
+}
+
+// Sorts the rows in the partitions of the table to be in primary key order.
+func (td *TableData) sortRows() {
+	type pkfield struct {
+		i int
+		c *sql.Column
+	}
+	var pk []pkfield
+	for _, column := range td.schema.Schema {
+		if column.PrimaryKey {
+			idx, col := td.getField(column.Name)
+			pk = append(pk, pkfield{idx, col})
+		}
+	}
+
+	less := func(l, r sql.Row) bool {
+		for _, f := range pk {
+			r, err := f.c.Type.Compare(l[f.i], r[f.i])
+			if err != nil {
+				panic(err)
+			}
+			if r != 0 {
+				return r < 0
+			}
+		}
+		return false
+	}
+
+	var idx []partidx
+	for _, k := range td.partitionKeys {
+		p := td.partitions[string(k)]
+		for i := 0; i < len(p); i++ {
+			idx = append(idx, partidx{string(k), i})
+		}
+	}
+
+	sort.Sort(partitionssort{td.partitions, idx, less})
+}
+
+func insertValueInRows(ctx *sql.Context, data *TableData, colIdx int, colDefault *sql.ColumnDefaultValue) error {
```

Comment:
altering a row seems like the kind of thing we should consolidate between dolt/gms

Reply:
I'll do another pass now that table rewrites work